component/authenticator/src/main/java/org/wso2/carbon/identity/authenticator/emailotp/EmailOTPAuthenticator.java

```diff
@@ -340,18 +347,37 @@ protected void processAuthenticationResponse(HttpServletRequest request, HttpSer
             succeededAttempt = validateWithBackUpCodes(context, userToken, authenticatedUser);
         }
 
+        /* If the email address of the user is not saved in the user's profile below code will be extracting the validated
+        email address from the context and save it in user's profile. If saving failed it will save a property in the
+        message context. */
         if (succeededAttempt && isLocalUser) {
             String username = authenticatedUser.toFullQualifiedUsername();
+            String userEmail;
             try {
-                String userEmail = getEmailValueForUsername(username, context);
-                if (StringUtils.isBlank(userEmail)) {
-                    Object verifiedEmailObject = context.getProperty(EmailOTPAuthenticatorConstants.REQUESTED_USER_EMAIL);
-                    if (verifiedEmailObject != null) {
+                userEmail = getEmailValueForUsername(username, context);
+            } catch (EmailOTPException e) {
+                throw new AuthenticationFailedException("Failed to get the email claim for user " + username + " for tenant "
+                        + context.getTenantDomain(), e);
+            }
+
+            if (StringUtils.isBlank(userEmail)) {
+                log.debug("User profile doesn't contain the email address. Updating the verified email address for user " + username);
+                Object verifiedEmailObject = context.getProperty(EmailOTPAuthenticatorConstants.REQUESTED_USER_EMAIL);
+                if (verifiedEmailObject != null) {
+                    try {
                         updateEmailAddressForUsername(context, username);
+                    } catch (UserStoreClientException e) {
+                        context.setProperty(EmailOTPAuthenticatorConstants.EMAIL_UPDATE_FAILURE, "true");
+                        throw new AuthenticationFailedException("Email Claim Update Failed for user " + username, e.getCause());
```

Comment:
change to                          throw new AuthenticationFailedException("Email Claim Update Failed for user " + username, e);